### PR TITLE
capi: Add object sizes to input types

### PR DESCRIPTION
**Pull request by danielocfb**
In order to ensure some level of forward & backward compatibility, this change extends all our input types with a member holding the objects size. Users need to set this member to the size of the type. The library can then ignore unsupported fields (forward compatibility) or provide reasonable defaults for fields not provided (backward compatibility).

### CODE REVIEW

**Comment by danielocfb** [comment]
Well, let me add logic for checking for zero "extensions" as well...

**Comment by d-e-s-o** [comment]
This should be ready for review now. Can you please take a look @anakryiko ?

File: capi/build.rs [hunk 1]
```diff
@@ -93,28 +126,29 @@ fn compile(compiler: &str, src: &Path, dst: &str, options: &[&str]) {
             .chain([src.as_os_str(), "-o".as_ref(), dst.as_os_str()]),
     )
     .unwrap_or_else(|err| panic!("failed to run `{compiler}`: {err}"));
+
+    let () = adjust_mtime(&dst).unwrap();
```
Comment (anakryiko) [inline capi/build.rs:130]: maybe split this into a separate commit and add a comment on why we are doing this?

File: capi/cbindgen.toml [hunk 2]
```diff
@@ -4,6 +4,29 @@ language = "C"
...
+ * This dance with uninitialized declaration, followed by memset to zero,
+ * followed by assignment using compound literal syntax is done to preserve
+ * ability to use a nice struct field initialization syntax and **hopefully**
+ * have all the padding bytes initialized to zero. It's not guaranteed though,
+ * when copying literal, that compiler won't copy garbage in literal's padding
+ * bytes, but that's the best way I've found and it seems to work in practice.
+ *
+ * Macro declares opts struct of given type and name, zero-initializes,
+ * including any extra padding, it with memset() and then assigns initial
+ * values provided by users in struct initializer-syntax as varargs.
```
Comment (anakryiko) [inline capi/cbindgen.toml:19]: let's leave a pointer to libbpf source code where you copied this from, both for attribution and blame distribution ;)

File: capi/cbindgen.toml [hunk 3]
```diff
@@ -4,6 +4,29 @@ language = "C"
...
+ */
+#define BLAZE_INPUT(TYPE, NAME, ...)        \\
+  struct TYPE NAME = ({                     \\
+    memset(&NAME, 0, sizeof(struct TYPE));  \\
+    (struct TYPE) {                         \\
+      ._size = sizeof(struct TYPE),         \\
```
Comment (anakryiko) [inline capi/cbindgen.toml:25]: very subjective, but `_size` doesn't look great due to underscore, tbh. I suspect `size` is painful to use from Rust side, right? Do you hate `sz`?

File: capi/cbindgen.toml [hunk 4]
```diff
@@ -38,6 +38,9 @@ exclude = ["Addr"]
 "Normalizer" = "struct blaze_normalizer"
 "Symbolizer" = "struct blaze_symbolizer"
 
+[layout]
+packed = "__attribute__((packed))"
```
Comment (anakryiko) [inline capi/cbindgen.toml:42]: oh, I wouldn't do packed... (I've been told) it forces compiler to generate less optimal code when accessing members of the struct (due to assumption that it can be at any unaligned address).

Let's drop packed, if we want to make sure we don't have any padding, we can use explicit "do-not-use-or-reserved" fields to make sure structs are zero initialized properly

File: capi/src/inspect.rs [hunk 5]
```diff
@@ -267,10 +276,14 @@ pub unsafe extern "C" fn blaze_inspect_syms_elf(
...
 ) -> *const *const blaze_sym_info {
+    if !input_zeroed!(src, blaze_inspect_elf_src) {
+        return ptr::null()
+    }
+
     // SAFETY: The caller ensures that the pointer is valid.
     let inspector = unsafe { &*inspector };
     // SAFETY: The caller ensures that the pointer is valid.
-    let src = Source::Elf(Elf::from(unsafe { &*src }));
+    let src = Source::Elf(Elf::from(unsafe { src.read_unaligned() }));
```
Comment (anakryiko) [inline capi/src/inspect.rs:290]: read_unaligned() will just read `sizeof(struct blaze_inspect_elf_src)` bytes, right? If yes, then it's problematic, because if user was compiled against older version of the library, they provided not enough bytes, so you'd be reading out of bounds.

On libbpf side I have OPTS_GET() macro, which allows to fetch information field-by-field, using default value, if user didn't provide enough bytes.

An alternative to this is to do what kernel does. Have a local full struct, memset(0) it, copy however many bytes user provided over local struct variable, and then use local variable, knowing that all extra fields are zero. This simplifies code a lot at the price of a tiny mem copy overhead. But you can actually structure code to avoid this overhead if user provided exactly the right amount of bytes (or even more).

File: capi/include/blazesym.h [hunk 6]
```diff
@@ -353,6 +380,10 @@ typedef struct blaze_symbolize_inlined_fn {
    * Source code location information for the inlined function.
    */
   struct blaze_symbolize_code_info code_info;
+  /**
+   * Unused member available for future expansion.
+   */
+  uint8_t _reserved[8];
```
Comment (anakryiko) [inline capi/include/blazesym.h:386]: do you support (plan to support) 32-bit build of blazesym? Because that will change pointer sizes in these structs and so we'd have to think about amount of reserved bytes needed

File: capi/include/blazesym.h [hunk 7]
```diff
@@ -49,7 +53,10 @@ typedef enum blaze_sym_type {
    * The symbol is a variable.
    */
   BLAZE_SYM_VAR,
-} blaze_sym_type;
+};
+#ifndef __cplusplus
+typedef uint8_t blaze_sym_type;
```
Comment (anakryiko) [inline capi/include/blazesym.h:58]: it's a pretty clever abuse of C++ not requiring `enum` prefix :)

I think alternative would be to use packed enum fields (e.g., https://takashiidobe.com/gen/pack-enums-in-c), which would use enum consistently.

WDYT?

File: capi/include/blazesym.h (hunk 6 above)
Comment (d-e-s-o) [reply capi/include/blazesym.h:386]: I don't see a reason not to support it. Why does it matter, though? We reserve a bit more, percentage-wise, perhaps to something not an 8-byte boundary, but what impact does it have?

File: capi/src/inspect.rs (hunk 5 above)
Comment (d-e-s-o) [reply capi/src/inspect.rs:290]: Yep. I'd say it's fine; once we add additional fields we need to cope with it. This version of the library will never have additional fields, and none will ever have fewer -- so there cannot be any out of bounds reads. Only once we add a field we need to make sure to make it optional, I think.

File: capi/cbindgen.toml (hunk 3 above)
Comment (d-e-s-o) [reply capi/cbindgen.toml:25]: I wouldn't say it's painful to use, it just overlaps with the `size` member of `blaze_sym_info`. But yeah, not a fan of `sz`. If anything it makes it more confusing from my point of view, because while the underscore is kind of sort of meant to indicate irrelevance to the user (they use a macro), `sz` is just an abbreviated `size`.

I suppose we could rename `blaze_sym_info.size` to `blaze_sym_info.sym_size` and then go with `size`, but not sure if that fundamentally changes much for the better. Or we could call `_size` `type_size` or something like that.

File: capi/include/blazesym.h (hunk 7 above)
Comment (d-e-s-o) [reply capi/include/blazesym.h:58]: This is `cbindgen` generated, based on `repr(u8)`. I don't believe there is an option to force this attribute, because `repr(packed)` in Rust only works on `struct` and `union`, not `enum`.

File: capi/cbindgen.toml (hunk 2 above)
Comment (d-e-s-o) [reply capi/cbindgen.toml:19]: Okay. I think the `memset` dance is no longer necessary with packed structs, I just forgot to remove it.

File: capi/cbindgen.toml (hunk 4 above)
Comment (d-e-s-o) [reply capi/cbindgen.toml:42]: I started off with non-packed and it's just a huge *()!#&*($!@#*()(#!*!)(:
- it requires this `_last_field` dance
- it requires `offset_of!` macro, which Rust doesn't currently have built-in (requires additional dependency; though it's slated to stabilize with 1.77, I think)
- I don't think it works reliably in Rust. We can zero initialize a struct from a zeroed byte buffer, but the moment we copy it (which happens way more often than in C, mind you) the compiler may fill padding/trailing bytes with arbitrary data, breaking the logic; I would not be surprised if the same is true for C, by the way, it's just that there copying a struct using built-in copy syntax is rarely ever done, and presumably never for the types that `libbpf` uses it for
  - yes, these types are meant for C code, but still, we need them in Rust for testing purposes

These aren't performance critical struct by any means. At least on x86, I'd think the generated code is likely the same, only that loads could potentially be penalized if unaligned. But again, the types that are packed are option types that are not in a performance critical path.

**Comment by d-e-s-o** [comment]
I addressed first set of comments (pertaining `BLAZE_INPUT` macro and separate commit for modified time stamp modification). Let's discuss the rest. Thanks for the review!

File: capi/cbindgen.toml (hunk 4 above)
Comment (d-e-s-o) [reply capi/cbindgen.toml:32]: I suppose making padding bytes explicit is an option. But it's still more error prone, I'd say, because it's easy to forget/get wrong and not easily automatically verifiable -- whereas with packed we know there is nothing else.
Let me know how strongly you feel.

File: capi/include/blazesym.h (hunk 6 above)
Comment (anakryiko) [reply capi/include/blazesym.h:376]: I was just thinking if the idea is to reserve a space one extra pointer, then the reserved size might be smaller. But yes, in the gran scheme of things extra few bytes shouldn't be a big deal.

File: capi/src/inspect.rs (hunk 5 above)
Comment (anakryiko) [reply capi/src/inspect.rs:290]: I guess I see no reason to not solve this upfront consistently for all use cases where you add size field, while all this is fresh. Technically, someone can just allocate 8 bytes for size field and pass such "noop" options, which wouldn't be illegal under this sized options contract, even though it's clearly a bit of an abuse.

Anyways, I think it's good to think and handle this upfront once and never have to think about this again, that's my only point.

File: capi/include/blazesym.h (hunk 6 above)
Comment (d-e-s-o) [reply capi/include/blazesym.h:376]: The problem is that in general we can't anticipate what will be added. If we just add space for a `usize`/pointer then we run into problems if we were to extend the type by, say, an `Addr`/`u64`.

File: capi/cbindgen.toml (hunk 4 above)
Comment (anakryiko) [reply capi/cbindgen.toml:32]: I just feel uneasy about this packed thing, that's all. I'm not saying it will ever cause problems, but my gut feel says it might, which is why I'm pushing back a bit. If you feel confident about this approach, go for it.

last_field isn't a big deal, you don't change that often and neither there are a lot of structs with size field, for offset_of(), until it stabilizes we can use extra crate of create our own version of offset_of() (not sure how involved it is, but there seems to be https://github.com/Gilnaa/memoffset/ available). Explicit padding at the end is indeed error-prone, especially taking into account 32-bit vs 64-bit architectures, but it's all doable.

Again, this is just a gut feel and I don't think I've seen many libraries that use packed structs on their "main" types.

re: performance, you might not penalize x86-64 much, but you probably can on arm64 (but again, I don't have much knowledge about arm64 specifics). Anyways, I'm not going to fight for this, but I'd probably not opt into packed struct myself.

File: capi/cbindgen.toml (hunk 3 above)
Comment (anakryiko) [reply capi/cbindgen.toml:15]: out of those options, I'd probably vote for renaming to sym_size (it makes sense), and use cleaner "size" field name. That underscore is just an eye sore, tbh.

Comment (d-e-s-o) [reply capi/cbindgen.toml:15]: I went with the `type_size` approach.

File: capi/cbindgen.toml (hunk 4 above)
Comment (d-e-s-o) [reply capi/cbindgen.toml:32]: I removed `packed` and made trailing padding explicit now.

**Comment by d-e-s-o** [comment]
Comments should have been addressed now. Can you please take another look when you get a chance, @anakryiko ?

**Review by anakryiko — APPROVED** [review]
alright, thanks! LGTM.
